### PR TITLE
provide Base.crc32c checksum and use it in checking cache staleness

### LINKED ISSUE #17845: Precompile should check staleness using a hash instead of a timestamp
As discussed in https://github.com/JuliaStats/StatsBase.jl/issues/202, alternating `Pkg.add(X)` and `using X` statements can lead to errors. According to @tkelman, this could be fixed by making `require` check staleness by looking at the timestamp instead of a hash.

Example:

```julia
$ rm -rf ~/.julia/v0.4
$ rm -rf ~/.julia/lib/v0.4
$ julia -e "Pkg.update()"
INFO: Initializing package repository /Users/tamasnagy/.julia/v0.4
INFO: Cloning METADATA from git://github.com/JuliaLang/METADATA.jl
INFO: Updating METADATA...
INFO: Computing changes...
INFO: No packages to install, update or remove
$ julia
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.4.5 (2016-03-18 00:58 UTC)
 _/ |\__'_|_|_|\__'_|  |
|__/                   |  x86_64-apple-darwin15.4.0

julia> Pkg.add("StatsBase")
INFO: Installing BinDeps v0.4.1
INFO: Installing Compat v0.8.6
INFO: Installing Rmath v0.1.2
INFO: Installing SHA v0.2.0
INFO: Installing StatsBase v0.9.0
INFO: Installing StatsFuns v0.3.0
INFO: Installing URIParser v0.1.5
INFO: Building Rmath
INFO: Package database updated

julia> using StatsBase
INFO: Precompiling module StatsBase...

julia> Pkg.add("Distributions")
INFO: Installing Calculus v0.1.15
INFO: Installing Distributions v0.10.0
INFO: Installing PDMats v0.4.2
INFO: Building Rmath
INFO: Package database updated

julia> using Distributions
INFO: Precompiling module Distributions...
INFO: Recompiling stale cache file /Users/tamasnagy/.julia/lib/v0.4/Rmath.ji for module Rmath.
INFO: Recompiling stale cache file /Users/tamasnagy/.julia/lib/v0.4/StatsFuns.ji for module StatsFuns.
INFO: Recompiling stale cache file /Users/tamasnagy/.julia/lib/v0.4/StatsBase.ji for module StatsBase.
INFO: Recompiling stale cache file /Users/tamasnagy/.julia/lib/v0.4/Distributions.ji for module Distributions.
WARNING: Module Rmath uuid did not match cache file
  This is likely because module Rmath does not support  precompilation but is imported by a module that does.
ERROR: __precompile__(true) but require failed to create a precompiled cache file
 in require at /usr/local/Cellar/julia/0.4.5/lib/julia/sys.dylib
```

### PR DESCRIPTION
**Pull request by stevengj**
This fixes #17845: to reduce the chance of false positives when checking cache staleness, it also checks a CRC-32c checksum of the file if the timestamp doesn't match.  (That is, the file is considered stale if `timestamp mismatch && checksum mismatch`).

The performance impact seems to be negligible.  In the common case where the timestamps match, the checksum is never computed at all, and in any case the computation of a checksum seems to be orders of magnitude faster than the time to load the cache file.

I used Mark Adler's (@madler) ~~BSD-licensed~~ [zlib-licensed](https://opensource.org/licenses/Zlib) CRC-32c implementation that he [posted on StackOverflow](http://stackoverflow.com/questions/17645167/implementing-sse-4-2s-crc32c-in-software), which includes both software and hardware-accelerated versions (the latter for SSE4.2 hardware).   [Adler](https://en.wikipedia.org/wiki/Mark_Adler) seems to be someone who knows what he is doing with checksum algorithms, so I was happy to see he posted usable code.  His code seems to be pretty fast (well over an order of magnitude faster than @andrewcooke's implementation in CRC.jl, which Andrew said is comparable to zlib; see also andrewcooke/CRC.jl#5).  @baruch did [some benchmarking](https://github.com/baruch/crcbench) and he found that Adler's implementation was one of the fastest.   I don't see much point in re-implementing it in Julia.

Because of the existence of hardware support, [CRC-32c seems like the most sensible choice](http://www.evanjones.ca/crc32c.html) of checksum algorithm, so I didn't really look at alternatives.

It also seems like a good idea to export the `crc32c` checksum function, since this is pretty generally useful functionality to have, but I wanted to get feedback first.

### CODE REVIEW

**Comment by tkelman** [comment]
should be listed in the license file and exception list in contrib/add_license_to_files.jl.

File: src/support/crc32c.c [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+    /* populate shift tables the first time through */
+    /* pthread_once(&crc32c_once_hw, crc32c_init_hw); */
+
+    /* pre-process the crc */
+    crc0 = crc ^ 0xffffffff;
+
+    /* compute the crc for up to seven leading bytes to bring the data pointer
+       to an eight-byte boundary */
+    while (len && ((uintptr_t)next & 7) != 0) {
+        __asm__("crc32b\t" "(%1), %0"
```
Comment (tkelman) [inline src/support/crc32c.c]: should fall back to the software version for `_COMPILER_MICROSOFT_`

File: src/support/crc32c.c [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+                len -= 8;
+    }
+    while (len) {
+        crc = crc32c_table[0][(crc ^ *next++) & 0xff] ^ (crc >> 8);
+        len--;
+    }
+    return (uint32_t)crc ^ 0xffffffff;
+}
+
+#ifdef __x86_64__
```
Comment (tkelman) [inline src/support/crc32c.c]: would the instructions be available for 32 bit i686 binaries?

File: src/support/crc32c.c (hunk 1 above)
Comment (stevengj) [inline src/support/crc32c.c]: Done.

**Comment by stevengj** [comment]
Fixed the license file and updated the tests.

File: src/support/crc32c.c (hunk 2 above)
Comment (stevengj) [inline src/support/crc32c.c]: Yeah, probably.  It wouldn't hurt to enable it there, in any case, as the `cpuid` check for SSE4.2 will work in 32-bit mode.

Comment (stevengj) [inline src/support/crc32c.c]: Fixed.

File: test/compile.jl [hunk 3]
```diff
@@ -159,16 +159,23 @@ try
           """)
 
     Base.compilecache("FooBar")
-    sleep(2)
     @test isfile(joinpath(dir, "FooBar.ji"))
 
+    sleep(2) # wait to make sure file modification date changes
```
Comment (tkelman) [inline test/compile.jl]: you might need a bit more than 2 seconds to ensure the *date* changes

(not serious)

**Comment by stevengj** [comment]
We could use [__cpuid](https://msdn.microsoft.com/en-us/library/hskdteyh.aspx) and [_mm_crc32_u64](https://msdn.microsoft.com/en-us/library/bb514033(v=vs.120).aspx) with MS compilers, rather than inline asm.

File: src/support/crc32c.c [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+#endif
+        crc32c_init_sw();
+}
+
+/* Compute a CRC-32C.  If the crc32 instruction is available, use the hardware
+   version.  Otherwise, use the software version. */
+JL_DLLEXPORT uint32_t jl_crc32c(uint32_t crc, const void *buf, size_t len)
+{
+    return
+#ifdef __x86_64__
```
Comment (tkelman) [inline src/support/crc32c.c]: HW_CRC ?

**Comment by stevengj** [comment]
(Speaking of future optimizations, ARMv8 also has a `CRC32CX` instruction for CRC-32c of 8 bytes.)

File: src/support/crc32c.c (hunk 2 above)
Comment (stevengj) [inline src/support/crc32c.c]: Hmm, clang cc doesn't seem to like the crc instructions in 32-bit mode, even if I compile with `-msse4.2`.   I've disabled them for now in 32-bit mode; we can always add them back later if we figure out how/whether it is possible.

**Comment by stevengj** [comment]
Yay, tests are green.

File: src/support/crc32c.c (hunk 2 above)
Comment (tkelman) [inline src/support/crc32c.c]: where are we using clang that's 32 bit?

File: LICENSE.md [hunk 5]
```diff
@@ -29,6 +29,7 @@ for exceptions.
 
 Julia includes code from the following projects, which have their own licenses:
 
+- [crc32c.c](http://stackoverflow.com/questions/17645167/implementing-sse-4-2s-crc32c-in-software) (CRC-32c checksum code by Mark Adler) [[ZLib](https://opensource.org/licenses/Zlib)].
```
Comment (Keno) [inline LICENSE.md]: Missing newline and `-`?

File: src/support/crc32c.c (hunk 2 above)
Comment (Keno) [inline src/support/crc32c.c]: I believe the `msse` flags only affect the intrinsics, while the inline assembly is controlled by the cpu/arch flags.

File: LICENSE.md (hunk 5 above)
Comment (stevengj) [inline LICENSE.md]: No, the line has a newline and a `-`.  Maybe the diff format is confusing?

Comment (Keno) [inline LICENSE.md]: Yes, sorry, I'm gonna go clean my glasses.

**Comment by stevengj** [comment]
@StefanKarpinski, @vtjnash, any comment?

**Comment by vtjnash** [comment]
I'm still against using a checksum to validate files. Remember too that with precompile, you will pay this cost m*n^2 times, so even a small cost quickly adds up and requires moving many megabytes of extra data which may not be local (the precompile .ji files are all supposed to be local)

**Comment by stevengj** [comment]
@vtnash, the files whose timestamps don't match but whose checksums match will typically be small `deps.jl` files re-generated by `Pkg.build`.   There is no way that the time to checksum those, even if they are on a remote server, will be slower than the time to load a module from the cache.

The number of times that this occurs is irrelevant, because in the cases where you run the same checksum a large number of times you are also loading the module from a cache a large number of times.

**Comment by vtjnash** [comment]
We can easily fix that bug in BinDeps rather than working around it here. 

I don't see any reason loading a module should be slower than doing CRC over the sources. In fact, I think that loading the module should be less work.

I don't find it terribly convincing that this only affects a small number of cases, for me that's also a negative. 

Another very strong negative, IMO, is that doing touch or save in my editor won't trigger a rebuild. With this change, I now need to know where to find a hidden cache file and delete it if something goes wonky. That violates the goal of the cache being fully transparent.

**Comment by tkelman** [comment]
Changing the timestamp without changing the content *shouldn't* trigger a rebuild. Timestamps aren't a great proxy for matching content. This will make .ji files more portable and easier to deploy as part of prebuilding Julia along with sets of packages.

**Comment by vtjnash** [comment]
Timestamps aren't simply a proxy for content, they are part of the content, and one that is easy for the user to influence. In that regard, I could perhaps accept `stale = (same-file-path | same-crc) && same-time-stamp`. But if you're going to make a claim to the contrary, you at least need to give some supporting rationale 

Deployment is a different process with different requirements. I don't see any reason to couple it to the behavior of precompile.

**Comment by tkelman** [comment]
> if you're going to make a claim to the contrary, you at least need to give some supporting rationale

Because the timestamp of a Julia file doesn't change the end-result behavior of the code contained in it whatsoever?

**Comment by vtjnash** [comment]
Hah. That assumes that the compiler is pure with respect to the universe, which is nonsense. It's not even best effort; I would give it a grade of: "tries to be fairly predictable if everything goes well". But fortunately a changed timestamp is a strong indicator that the user did intend to change something, so we can mostly recover from the inaccuracy of the rest of the system.

**Comment by tkelman** [comment]
`include` doesn't change behavior depending on the timestamp of the file, except as an implementation detail of the current staleness check. If we're arguing about "strong indicator that the user did intend to change something" then we're trying to read minds, and so far you're the only person arguing on the "timestamps should matter" side vs multiple people who've asked for them not to be the sole criterion - https://github.com/JuliaLang/julia/issues/17845#issuecomment-237934105.

**Comment by vtjnash** [comment]
BinDeps set the "this dependency has changed and needs to be rebuilt" flag. It's not the fault of base that we honor this flag, nor the fault of the original author (since the capability didn't exist at the time). But seriously, fix BinDeps.

I'm not trying to read minds, because I don't have to. I can simply read the filesystem metadata that has been the standard for build systems, noting that the frequent failure there is that it is not strict enough, and arguing that we shouldn't react by making our heuristics weaker. Nobody ever wished that it might not be possible to get `make` to recompile a C file after changing it. Instead, you re-`touch` the source file and let the build system take care of propagating that to the result.

`include` also rebuilds the whole cache. You've lost me on how this is an argument for not rebuilding the cache.

**Comment by tkelman** [comment]
https://github.com/JuliaLang/julia/issues/17845#issuecomment-237934105 wasn't about BinDeps.

You're equating filesystem metadata with "indicator that the user did intend to change something," when they're often not connected at all.

> Nobody ever wished that it might not be possible to get `make` to recompile a C file after changing it.

The existence of `ccache` disagrees with you.

**Comment by vtjnash** [comment]
> when they're often not connected at all.

Rebuilding a false positive is fine as long as it doesn't get stuck repeatedly attempting to the update. It is not acceptable for a build system to have "not often" false negatives which have no easy recourse for the user to fix them.

> The existence of ccache disagrees with you.

not at all, it agrees with me perfectly. First of all, `ccache` is a compiler, not a make dependency system. Also, it's a bug if `ccache` output isn't perfectly captured by the input (https://ccache.samba.org/manual.html#_bugs), and it includes rigorous checks that the inputs are the same, such as requiring that the filenames are exactly the same and parsing the source to prove it doesn't contain any content that could result in a detectable difference in output.

I actually intend to write a ccache-like too for Julia, although designed around Julia's semantics so it will be completely different.

**Comment by stevengj** [comment]
Okay, I'm convinced.

**Comment by StefanKarpinski** [comment]
> I don't see any reason loading a module should be slower than doing CRC over the sources. In fact, I think that loading the module should be less work.

I have to call b.s. on this statement. Are you really claiming that reading, parsing and compiling a Julia file will take less time than computing a CRC on the file?

**Comment by vtjnash** [comment]
no, I'm claiming that doing a crc on the source would add a significant percentage to the time to load the precompile file, so it would add a bit of unpredictable uncertainty to the load time. In other-words, I'm committing to making precompile loading much faster :)

**Comment by vtjnash** [comment]
fwiw, please don't delete this branch, as having CRC code is likely to be useful (for example, we could be using it to validate that the .ji files aren't truncated or corrupted.)

**Comment by stevengj** [comment]
@vtjnash, yes, that was my thought.

**Comment by stevengj** [comment]
Maybe I should open another PR with just the CRC?

**Comment by vtjnash** [comment]
sgtm. If you didn't, I was planning on doing it myself. Probably not a huge cost, but I would prefer – flr the sake of startup time – to either make the table data fully static, or lazy initialize it (using `uv_once`), with preference for the former as it allows the kernel to share these data pages between processes.

**Pull request closed without merging**